File: kafka-0.7.0/kafka-0.7.0-core/src/main/scala/kafka/server/KafkaRequestHandlers.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * 
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.server

import org.apache.log4j.Logger
import kafka.log._
import kafka.network._
import kafka.message._
import kafka.api._
import kafka.common.ErrorMapping
import kafka.utils.{Utils, SystemTime}
import java.io.IOException

/**
 * Logic to handle the various Kafka requests
 */
private[kafka] class KafkaRequestHandlers(val logManager: LogManager) {
  
  private val logger = Logger.getLogger(classOf[KafkaRequestHandlers])
  private val requestLogger = Logger.getLogger("kafka.request.logger")
  println("-------------KafkaRequestHandlers-------------------init---------------------")
  def handlerFor(requestTypeId: Short, request: Receive): Handler.Handler = {
    println("-------------KafkaRequestHandlers-------------------handlerFor---------------------requestTypeId=" + requestTypeId)
    requestTypeId match {
      case RequestKeys.Produce => handleProducerRequest _
      case RequestKeys.Fetch => handleFetchRequest _
      case RequestKeys.MultiFetch => handleMultiFetchRequest _
      case RequestKeys.MultiProduce => handleMultiProducerRequest _
      case RequestKeys.Offsets => handleOffsetRequest _
      case _ => throw new IllegalStateException("No mapping found for handler id " + requestTypeId)
    }
  }
  
  def handleProducerRequest(receive: Receive): Option[Send] = {
    println("-------------KafkaRequestHandlers-------------------handleProducerRequest--------处理发送请求-----start--------------")
    val sTime = SystemTime.milliseconds
    val request = ProducerRequest.readFrom(receive.buffer)// 从收到的buffer 中读取数据

    if(requestLogger.isTraceEnabled){
      requestLogger.trace("Producer request " + request.toString)
    }
    handleProducerRequest(request, "ProduceRequest")
    if (logger.isDebugEnabled){
      logger.debug("kafka produce time " + (SystemTime.milliseconds - sTime) + " ms")
    }
    println("-------------KafkaRequestHandlers-------------------handleProducerRequest--------处理发送请求-----end--------------")
    None
  }

  def handleMultiProducerRequest(receive: Receive): Option[Send] = {
    println("-------------KafkaRequestHandlers-------------------handleMultiProducerRequest---------------------------")
    val request = MultiProducerRequest.readFrom(receive.buffer)
    if(requestLogger.isTraceEnabled)
      requestLogger.trace("Multiproducer request " + request.toString)
    request.produces.map(handleProducerRequest(_, "MultiProducerRequest"))
    None
  }

  private def handleProducerRequest(request: ProducerRequest, requestHandlerName: String) = {
    println("-------------KafkaRequestHandlers-------------------handleProducerRequest--------ProduceRequest--------2-----------")
    val partition = request.getTranslatedPartition(logManager.chooseRandomPartition) //选择需要数据写入的分区
    try {
      logManager.getOrCreateLog(request.topic, partition).append(request.messages) // 需要根据topic 和 partition，创建日志文件写入数据，如果存在，或者追加消息
      if(logger.isTraceEnabled)
        logger.trace(request.messages.sizeInBytes + " bytes written to logs.")
    }
    catch {
      case e =>
        logger.error("Error processing " + requestHandlerName + " on " + request.topic + ":" + partition, e)
        e match {
          case _: IOException =>
            logger.fatal("Halting due to unrecoverable I/O error while handling producer request: " + e.getMessage, e)
            Runtime.getRuntime.halt(1)
          case _ =>
        }
        throw e
    }
    None
  }

  def handleFetchRequest(request: Receive): Option[Send] = {
    val fetchRequest = FetchRequest.readFrom(request.buffer)
    if(requestLogger.isTraceEnabled)
      requestLogger.trace("Fetch request " + fetchRequest.toString)
    Some(readMessageSet(fetchRequest))
  }
  
  def handleMultiFetchRequest(request: Receive): Option[Send] = {
    val multiFetchRequest = MultiFetchRequest.readFrom(request.buffer)
    if(requestLogger.isTraceEnabled)
      requestLogger.trace("Multifetch request")
    multiFetchRequest.fetches.foreach(req => requestLogger.trace(req.toString))
    var responses = multiFetchRequest.fetches.map(fetch =>
        readMessageSet(fetch)).toList
    
    Some(new MultiMessageSetSend(responses))
  }

  private def readMessageSet(fetchRequest: FetchRequest): MessageSetSend = {
    var  response: MessageSetSend = null
    try {
      logger.trace("Fetching log segment for topic = " + fetchRequest.topic + " and partition = " + fetchRequest.partition)
      val log = logManager.getOrCreateLog(fetchRequest.topic, fetchRequest.partition)
      response = new MessageSetSend(log.read(fetchRequest.offset, fetchRequest.maxSize))
    }
    catch {
      case e =>
        logger.error("error when processing request " + fetchRequest, e)
        response=new MessageSetSend(MessageSet.Empty, ErrorMapping.codeFor(e.getClass.asInstanceOf[Class[Throwable]]))
    }
    response
  }

  def handleOffsetRequest(request: Receive): Option[Send] = {
    val offsetRequest = OffsetRequest.readFrom(request.buffer)
    if(requestLogger.isTraceEnabled)
      requestLogger.trace("Offset request " + offsetRequest.toString)
    val log = logManager.getOrCreateLog(offsetRequest.topic, offsetRequest.partition)
    val offsets = log.getOffsetsBefore(offsetRequest)
    val response = new OffsetArraySend(offsets)
    Some(response)
  }
}

File: kafka-0.7.0/kafka-0.7.0-core-java/src/main/java/kafka/server/KafkaRequestHandlers.java
package kafka.server;

import kafka.api.*;
import kafka.common.ErrorMapping;
import kafka.log.Log;
import kafka.log.LogManager;
import kafka.message.MessageSet;
import kafka.network.Receive;
import kafka.network.Send;
import kafka.utils.SystemTime;
import org.apache.log4j.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class KafkaRequestHandlers {
    private Logger logger = Logger.getLogger(KafkaRequestHandlers.class);
    private Logger requestLogger = Logger.getLogger("kafka.request.logger");
    private LogManager logManager;
    
    public KafkaRequestHandlers(LogManager logManager) {
        this.logManager = logManager;
    }
    
    public Send handlerFor(Short requestTypeId, Receive receive) {
        Send send = null;
        if (requestTypeId == RequestKeys.Produce) {
            send = handleProducerRequest(receive);
        } else if (requestTypeId == RequestKeys.Fetch) {
            send = handleFetchRequest(receive);
        } else if (requestTypeId == RequestKeys.MultiFetch) {
            send = handleMultiFetchRequest(receive);
        } else if (requestTypeId == RequestKeys.MultiProduce) {
            send = handleMultiProducerRequest(receive);
        } else if (requestTypeId == RequestKeys.Offsets) {
            send = handleOffsetRequest(receive);
        } else {
            throw new IllegalStateException("No mapping found for handler id " + requestTypeId);
        }
        return send;
    }
    
    /**
     * 处理 根据offset 消息消息请求
     *
     * @param request
     * @return
     */
    Send handleOffsetRequest(Receive request) {
        OffsetRequest offsetRequest = OffsetRequest.readFrom(request.buffer());
        if (requestLogger.isTraceEnabled())
            requestLogger.trace("Offset request " + offsetRequest.toString());
        Log log = logManager.getOrCreateLog(offsetRequest.topic, offsetRequest.partition);
        Long[] offsets = log.getOffsetsBefore(offsetRequest);
        OffsetArraySend response = new OffsetArraySend(offsets);
        return response;
    }
    
    
    Send handleMultiProducerRequest(Receive receive) {
        MultiProducerRequest request = MultiProducerRequest.readFrom(receive.buffer());
        if (requestLogger.isTraceEnabled()) {
            requestLogger.trace("Multiproducer request " + request.toString());
        }
        
        for (ProducerRequest producerRequest : request.produces) {
            handleProducerRequest(producerRequest, "MultiProducerRequest");
        }
        return null;
    }
    
    
    Send handleMultiFetchRequest(Receive request) {
        MultiFetchRequest multiFetchRequest = MultiFetchRequest.readFrom(request.buffer());
        if (requestLogger.isTraceEnabled()) {
            requestLogger.trace("Multifetch request");
        }
        FetchRequest fetches[] = multiFetchRequest.fetches;
        
        for (FetchRequest req : fetches) {
            requestLogger.trace(req.toString());
        }
        List<MessageSetSend> responses = new ArrayList<>();
        
        for (FetchRequest fetch : fetches) {
            responses.add(readMessageSet(fetch));
        }
        return new MultiMessageSetSend(responses);
    }
    
    
    Send handleFetchRequest(Receive request) {
        FetchRequest fetchRequest = FetchRequest.readFrom(request.buffer());
        if (requestLogger.isTraceEnabled()) {
            requestLogger.trace("Fetch request " + fetchRequest.toString());
        }
        return readMessageSet(fetchRequest);
    }
    
    /**
     * 处理生产者请求
     *
     * @param receive
     * @return
     */
    public Send handleProducerRequest(Receive receive) {
        Long sTime = System.currentTimeMillis();
        ProducerRequest request = ProducerRequest.readFrom(receive.buffer());
        
        if (requestLogger.isTraceEnabled()) {
            requestLogger.trace("Producer request " + request.toString());
        }
        handleProducerRequest(request, "ProduceRequest");
        if (logger.isDebugEnabled()) {
            logger.debug("kafka produce time " + (SystemTime.getMilliseconds() - sTime) + " ms");
        }
        return null;
    }
    
    private void handleProducerRequest(ProducerRequest request, String requestHandlerName) {
        int randomP = logManager.chooseRandomPartition(request.topic);
        int partition = request.getTranslatedPartition(randomP);
        try {
            logManager.getOrCreateLog(request.topic, partition).append(request.messages);
            if (logger.isTraceEnabled())
                logger.trace(request.messages.sizeInBytes() + " bytes written to logs.");
        } catch (Exception e) {
            logger.error("Error processing " + requestHandlerName + " on " + request.topic + ":" + partition, e);
            logger.fatal("Halting due to unrecoverable I/O error while handling producer request: " + e.getMessage(), e);
            Runtime.getRuntime().halt(1);
        }
    }
    
    private MessageSetSend readMessageSet(FetchRequest fetchRequest) {
        MessageSetSend response = null;
        try {
            System.out.println("消费：Fetching log segment for topic = " + fetchRequest.topic + " and partition = " + fetchRequest.partition);
            logger.trace("Fetching log segment for topic = " + fetchRequest.topic + " and partition = " + fetchRequest.partition);
            Log log = logManager.getOrCreateLog(fetchRequest.topic, fetchRequest.partition);
            System.out.println("--------------------------MessageSetSend--------------log-----------" + log);
            MessageSet messageSet = log.read(fetchRequest.offset, (long) fetchRequest.maxSize);
            response = new MessageSetSend(messageSet);
        } catch (Exception e) {
            e.printStackTrace();
            logger.error("error when processing request " + fetchRequest, e);
            try {
                response = new MessageSetSend(MessageSet.Empty, ErrorMapping.codeFor(e.getClass().newInstance().getCause()));
            } catch (InstantiationException e1) {
                e1.printStackTrace();
            } catch (IllegalAccessException e1) {
                e1.printStackTrace();
            }
        }
        return response;
    }
    
    
}
